aten/src/ATen/core/boxing/impl/boxing.h

```diff
@@ -68,61 +103,195 @@ inline bool boxArgumentsOrCannotBoxIntoStack(std::vector<c10::IValue>& stack, co
+// particular, not all types appearing in op signatures are supported,
+// and ops returning references have nonstandard wrapper implementations.
+//
 
-template<class Result>
-std::enable_if_t<supports_boxing_result<Result>::value && !std::is_same<void, Result>::value, Result>
-callBoxedFunc(KernelFunction::InternalBoxedKernelFunction* boxed_kernel_func, OperatorKernel* functor, const OperatorHandle& opHandle, torch::jit::Stack& stack) {
-  (*boxed_kernel_func)(functor, opHandle, &stack);
-  TORCH_INTERNAL_ASSERT(stack.size() == 1, "A boxed kernel should only push one return to the stack");
-  return std::move(stack[0]).to<Result>();
-}
+// base definition should never be instantiated.
+// A "no call method defined on BoxedKernelWrapper" compile error means that
+// an op signature has failed to trigger any of the partial specializations
+// that follow.
+//
+template<class FuncType, class Enable = void>
+struct BoxedKernelWrapper {
+  static_assert(sizeof(FuncType) == -1,
+    "Function signature contains one or more unsupported parameter and/or return types. "
+    "Look for a nearby error like "
+    "\"'call' is not a member of 'c10::impl::BoxedKernelWrapper<(your function type), void>'\" "
+    "- (your function type) is the unsupported signature.");
+};
 
-template<class Result>
-std::enable_if_t<supports_boxing_result<Result>::value && std::is_same<void, Result>::value, Result>
-callBoxedFunc(KernelFunction::InternalBoxedKernelFunction* boxed_kernel_func, OperatorKernel* functor, const OperatorHandle& opHandle, torch::jit::Stack& stack) {
-  (*boxed_kernel_func)(functor, opHandle, &stack);
-  TORCH_INTERNAL_ASSERT(stack.size() == 0, "A boxed kernel returned a value but when we called it with KernelFunction::callBoxedFunc(), we expected it to return void.");
-}
+// 1. Unsupported type traps.
+//
+// These specializations capture the remaining gaps in boxing support.
+// Rather than triggering compile errors, we generate boxed kernels that
+// raise runtime errors. As support for these types is added, the
+// specializations can be removed.
+//
+
+// at::Dimname
+template <class... Args>
+using has_dimname_arg =
+  guts::disjunction<
+    std::is_same<at::Dimname, std::decay_t<Args>>...,
+    std::is_same<c10::ArrayRef<at::Dimname>, std::decay_t<Args>>...,
+    std::is_same<c10::optional<c10::ArrayRef<at::Dimname>>, std::decay_t<Args>>...
+  >;
 
 template<class Result, class... Args>
-Result boxAndCallBoxedFunc(KernelFunction::InternalBoxedKernelFunction* boxed_kernel_func, OperatorKernel* functor, const OperatorHandle& opHandle, Args... args, std::enable_if_t<!supports_boxing<Result, Args...>::value, int> = 0) {
-  // Some kernels don't need to actually box, and don't return.  If that's the
-  // case, just call them anyway without a stack.  These special cases can be
-  // removed once we support boxing everything.
-  // See Note [named_not_supported_kernel]
-  if (boxed_kernel_func == &named_not_supported_kernel) {
```

Comment:
@bhosmer I just realized that we lost this logic entirely. It is not a problem to lose this logic when we 100% support boxing, since it is obsolete at that point, but prior to that, missing this logic means that you'll get a much worse error message when you hit one of the special `named_not_supported_kernel` on a codepath that doesn't support boxing it. I don't recall how close to 100% boxing we are now so I don't know the severity of this.

Reply:
We're at 100% boxing now (mod reordering issue, but all types are supported) 😁  but thanks for the callout! I thought about this earlier today as well, can't remember what triggered it - something in your presentation maybe?